### ISSUE TITLE
Shaders not easy to compile on windows

### ISSUE THREAD

**Issue by tangmi** [servo/pathfinder#346]
I'm trying to modify the shaders in pathfinder (for the webgpu backend) on a Windows machine and am running into issues running the Makefile (not `pathfinder`'s fault, just the normal issues with running `make` on Windows, like missing unix tools, etc).

Would `pathfinder` consider letting this part of the pipeline be ported to Rust?

Some pros:
- It could include the required compilers (through `glslang_sys` and `spirv_cross` or something like `shaderc`) through a crate, so contributors don't need to install a compatible version on their system
- Could be configured as a `build.rs` script to automatically rebuild shaders if they change
- Hopefully will be more cross-platform than a Makefile (and more approachable to Rust folks who may not have Makefile experience!)

Cons:
- Would bloat the dependency size/build time for most contributors that do not need to modify shaders
- This is potentially over-engineering a solution to a problem that I'm the only one experiencing. :smiley:
- Edit: Also potentially less necessary when we have https://github.com/gfx-rs/naga

Thanks!

**Comment by s3bk** [comment]
This is unfortunately not just a windows problem. Compiling the tools is quite tricky and involves quite a few manual steps.

**Comment by tangmi** [comment]
The `build.rs` could use crates that do the clone/build of dependant repos for you (to `OUT_DIR`), like  `shaderc` (this still requires `cmake` and `ninja`, however). I have a build script like this working in my branch: https://github.com/tangmi/pathfinder/blob/23e23db0997db99a1ae34abe28f442416188a56e/shaders/build.rs using `shaderc` as the glsl -> spirv compiler and `spirv-cross` for cross-compilation.  (sample output also commited on that branch: e.g. [gl3/tile.fs.glsl](https://github.com/tangmi/pathfinder/blob/23e23db0997db99a1ae34abe28f442416188a56e/resources/shaders/gl3/tile.fs.glsl)). 
 
The background for this change is that I need SPIR-V shaders for the WebGPU backend (for separate samplers/textures, uniform blocks, etc), but I'm much more comfortable writing Rust than I am writing Makefiles... Happy to change it back to a Makefile (with guidance!) if my WebGPU backend is successful!

**Comment by pcwalton** [comment]
I generally compile under Bash on Windows/WSL. `glslangValidator` needs to be in your `$PATH`.

**Comment by tangmi** [comment]
I believe it also needs `spirv-cross` (which I've learned is available along with `glslangValidator` via the Vulkan SDK), `make`, and some various `coreutils` tools.

I've been able to build with bash on WSL by appending `.exe` to `glslangValidator` and `spirv-cross` in the Makefile (I haven't installed either binary within WSL itself).

I've also been able to build on Windows (without WSL) with these steps:
- Install the [Vulkan SDK](https://vulkan.lunarg.com/), ensuring the bin (e.g. `C:\VulkanSDK\1.2.135.0\Bin\`) folder in in my `$PATH`.
- Install `busybox` (I did so with [`scoop`](https://scoop.sh/): `scoop install busybox`)
- Run `ash -c 'make --jobs all'` or `sh -c 'make --jobs all'`

This does change the line endings , but otherwise results in the same shaders.

I will note that I did not notice `./resources/shaders/README.md` until now, instead expecting it to be at `./shaders/README.md`.  I'm happy to close this issue, since the docs do exist, but I'm also happy to add some docs detailing my Windows steps.
